pyverbs/wr.pyx

```diff
@@ -281,6 +285,58 @@ cdef class SendWR(PyverbsCM):
         """
         self.send_wr.qp_type.xrc.remote_srqn = remote_srqn
 
+    def set_tso(self, TSO tso not None):
+        """
+        Set the members of the tso struct in the send_wr's anonymous union.
+        :param tso: A TSO object to copy to the work request
+        :return: None
+        """
+        self.send_wr.tso.hdr = tso.buf
+        self.send_wr.tso.hdr_sz = tso.length
+        self.send_wr.tso.mss = tso.mss
+
+
+cdef class TSO(PyverbsCM):
+    """ Represents the 'tso' anonymous struct inside a send WR """
+    def __init__(self, data, length, mss):
+        super().__init__()
+        self.alloc_buf(data, length)
+        self.length = length
+        self.mss = mss
+
+    cpdef alloc_buf(self, data, length):
+        if self.buf != NULL:
+            free(self.buf)
```

Comment:
hm... I'm not sure if I'm convinced to move the free out ?
TSO class allocates this buffer in the class initialization, and if the user needs to use a new data he/she can simply call alloc_buf, and it will reallocate the buffer, instead of calling "free_buf" and then "alloc_buf" (as a C style) each time.
In case of moving free out, it would be easier to the user to create a new TSO object each time...
What do you think ?

Reply:
Actually, after taking a deeper look, it looks like more confusion now with the alloc_buf() name, because the buffer isn't only get allocated it also initialized with the data that passed to alloc_buf() and the length gets updated too, so the function isn't only allocating a buffer.

correct me if I'm wrong, I didn't see any use of the TSO class in this PR, so I'm wondering why you introduce it...

Reply:
Right, the function maybe should be renamed (as set_header for example, since it actually sets the TSO header).

Regarding the TSO class, it's not used in this PR.  WR (in general, and also the new post send) has a lot of attributes and options. Currently in this series there's no test that uses TSO, but was decided to add it in case someone adds a related feature and want to test it using pyverbs.

Reply:
> Regarding the TSO class, it's not used in this PR. WR (in general, and also the new post send) has a lot of attributes and options. Currently in this series there's no test that uses TSO, but was decided to add it in case someone adds a related feature and want to test it using pyverbs.

I have no idea who decide it, but please never do such thing like submitting dead code.
Please or delete the code or provide test for the TSO.

Reply:
Removed the TSO related commit for now.